src/cryptographic_primitives/proofs/low_degree_exponent_interpolation.rs

```diff
@@ -0,0 +1,215 @@
+                if i != j && a1 == a2 {
+                    return Err(LdeiProveError::AlphaNotPairwiseDistinct);
+                }
+            }
+        }
+
+        let x: Vec<P> = g
+            .iter()
+            .zip(&alpha)
+            .map(|(g, a)| g.clone() * w.evaluate(a))
+            .collect();
+
+        let d = w.degree();
+        let u = Polynomial::<P>::sample(d);
+        let a: Vec<P> = g
+            .iter()
+            .zip(&alpha)
+            .map(|(g, a)| g.clone() * u.evaluate(a))
+            .collect();
+
+        let hash_input: Vec<&P> = g.iter().chain(&x).chain(&a).collect();
+        let e = H::create_hash_from_ge::<P>(hash_input.as_slice());
+
+        let z = &u - &(w * &e);
+
+        let statement = LdeiStatement {
+            alpha,
+            g,
+            x,
+            degree: d,
+        };
+        let proof = LdeiProof { a, e, z };
+
+        Ok((statement, proof))
+    }
+
+    /// Verifies correctness of a statement
+    ///
+    /// ## Protocol
+    ///
+    /// The verifier checks that `e = H(g1,...,gm,x1,...,xm,a1,...,am)`, that
+    /// `deg(z) ≤ d`, and that `a_i = g_i^z(αlpha_i) * x_i^e` for all i, and accepts if all of this is
+    /// true, otherwise rejects.
+    pub fn verify<H>(&self, statement: &LdeiStatement<P>) -> Result<(), ProofError>
+    where
+        H: Hash,
+    {
+        let hash_input: Vec<&P> = statement
+            .g
+            .iter()
+            .chain(&statement.x)
+            .chain(&self.a)
+            .collect();
+        let e = H::create_hash_from_ge::<P>(hash_input.as_slice());
+        if e != self.e {
+            return Err(ProofError);
+        }
+        if self.z.degree() > statement.degree {
+            return Err(ProofError);
+        }
+
+        let expected_a: Vec<_> = statement
+            .g
+            .iter()
+            .zip(&statement.alpha)
+            .zip(&statement.x)
+            .map(|((g, a), x)| g.clone() * self.z.evaluate(&a) + x.clone() * e.clone())
+            .collect();
+
+        if self.a == expected_a {
+            Ok(())
+        } else {
+            Err(ProofError)
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum LdeiProveError {
+    #[error("`alpha`s are not pairwise distinct")]
+    AlphaNotPairwiseDistinct,
+    #[error("alpha.len() != g.len()")]
+    AlphaLengthDoesntMatchG,
+}
+
+#[cfg(test)]
+mod tests {
+    use std::iter;
+
+    use crate::arithmetic::BigInt;
+    use crate::cryptographic_primitives::hashing::hash_sha256::HSha256;
+    use crate::elliptic::curves::traits::ECScalar;
+    use crate::test_for_all_curves;
+
+    use super::*;
+
+    test_for_all_curves!(correctly_proofs);
+    fn correctly_proofs<P>()
+    where
+        P: ECPoint + Clone + PartialEq,
+        P::Scalar: ECScalar + Clone + PartialEq,
+    {
+        let poly = Polynomial::<P>::sample(5);
+
+        let alpha: Vec<P::Scalar> = (1..=10).map(|i| ECScalar::from(&BigInt::from(i))).collect();
```

Comment:
are you sure the generators are not supposed to be picked at random such that there is no way to compute `x` such that `g_i = x g_j` ? 
for example see how we generate the generator vector in bulletproofs

Reply:
The line you commented doesn't choose generators — it chooses list of α (a list of pairwise distinct scalar from Setup phase of the proof). [Line below] chooses generators completely at random. I agree that correlation like `g_i = x g_j` could potentially lead to vulnerabilities 

[Line below]: https://github.com/ZenGo-X/curv/blob/8882430e0ef823a3e4ced3c3df0c1ce290fe2c71/src/cryptographic_primitives/proofs/low_degree_exponent_interpolation.rs#L165

Reply:
no idea why I picked that line :) 
so will you change it ?

Reply:
Ah I found what code you're referring to — [generate_random_point]. Are you ok with me moving this function to `curv::elliptic::curves:: generate_random_point` ?

[generate_random_point]: https://github.com/ZenGo-X/bulletproofs/blob/6bdb41ed670838a6da2672dfa10928d22a2167a3/src/proofs/range_proof.rs#L744